Source/Dafny/Compilers/ICompiler.cs

```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.ObjectModel;
+using System.IO;
+
+namespace Microsoft.Dafny;
+
+public interface ICompiler {
```

Comment:
This interface is so much nicer. :)

Here's an idea to consider for the very next PR (since it's not necessary here and the PR is already large): could we add a method to be invoked earlier in the pipeline, before verification, that can produce compiler-specific errors? Something like `CheckCompatibility`.

The use case would be a codebase using a specific compiler, especially one of the more incomplete ones like C++, where it's currently pretty frustrating to spend a ton of time getting your code to verify successfully only to then find out "oh wait, my compiler doesn't support unbounded `int`s" and have to start over. This is particularly easy for newer Dafny programmers to hit, since the documentation on which features are supported by which compilers is something you have to think to look up, and isn't that obvious for newcomers.

Reply:
> Here's an idea to consider for the very next PR (since it's not necessary here and the PR is already large): could we add a method to be invoked earlier in the pipeline, before verification, that can produce compiler-specific errors? Something like CheckCompatibility.

The compiler is fast and verification errors lead to runtime (not compile-time) errors, so we could just run the compiler first or in parallel with verification instead of running verification first and only then running the compiler.  So instead of `resolve → verify → translate → compile` we could just run 

```
resolve → verify
        → translate → compile
```

Or even `resolve → translate → verify → compile` if we don't want concurrency.

Reply:
Probably a good entry for your compiler planning doc!

Reply:
Yup, you're absolutely right that there's other ways to get the same result (and it would be better to not have another method to implement if possible). I'm not sure I believe "verification errors lead to runtime (not compile-time) errors" is always true, or that we want to ensure it is always true in the future, but we can work that into the design. :)